### PR TITLE
Xocl hw context support

### PR DESCRIPTION
**Pull request by saifuddin-xilinx**
Hw Context is in place. 
This PR is targeted to enable hw context changes for PCIe shim, XOCL driver, KDS core. 
This changes are align with XRT core hw context support. 

Currently enabling  hw_context default to verify all the regression test  cases are working fine or not. 
But hw_context  will be a added to xrt.ini option and set as false by default.  

Basic xbutil validate tests are working on vck5000 device.

### CODE REVIEW

File: src/runtime_src/core/common/config_reader.h [hunk 1]
```diff
@@ -552,6 +552,17 @@ get_ert_polling()
   return value;
 }
 
+/**
+ * Use new hw context for multi slot application 
+ */
+inline bool
+get_hw_context_flag()
+{
+  /**
```
Comment (rozumx) [inline src/runtime_src/core/common/config_reader.h:561]: Please remove empty comments.

File: src/runtime_src/core/common/drv/kds_core.c [hunk 2]
```diff
@@ -1076,13 +1081,53 @@ _kds_fini_client(struct kds_sched *kds, struct kds_client *client,
 	mutex_unlock(&client->lock);
 }
 
+
+static inline void
```
Comment (rozumx) [inline src/runtime_src/core/common/drv/kds_core.c:1085]: Does this function really need to be "inlined"?

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_ioctl.c [hunk 3]
```diff
@@ -91,6 +95,92 @@ int xocl_ctx_ioctl(struct drm_device *dev, void *data,
...
+ * for that process if loded into same slot.
+ */
+int xocl_create_hw_ctx_ioctl(struct drm_device *dev, void *data,
+        struct drm_file *filp)
+{
+	struct drm_xocl_create_hw_ctx *drm_hw_ctx =
+	       	(struct drm_xocl_create_hw_ctx *)data;
+        struct xocl_drm *drm_p = dev->dev_private;
+        struct xocl_dev *xdev = drm_p->xdev;
+        struct drm_xocl_axlf axlf_obj_ptr = { 0 };
```
Comment (rozumx) [inline src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_ioctl.c:111]:
```suggestion
        struct drm_xocl_axlf axlf_obj_ptr = { };
```

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c [hunk 4]
```diff
@@ -348,6 +356,232 @@ static int xocl_context_ioctl(struct xocl_dev *xdev, void *data,
...
+        cu_info->cu_idx = get_domain_idx(args->cu_index);
+}
+
+int xocl_open_cu_context(struct xocl_dev *xdev, struct drm_file *filp,
+		struct drm_xocl_open_cu_ctx *drm_cu_args)
+{
+	struct kds_client *client = filp->driver_priv;
+	struct kds_client_hw_ctx *hw_ctx = NULL;
+	struct kds_client_cu_ctx *cu_ctx = NULL;
+	struct kds_client_cu_info cu_info = { 0 }; // Local Variable
```
Comment (rozumx) [inline src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c:497]:
```suggestion
	struct kds_client_cu_info cu_info = {}; // Local Variable
```

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c [hunk 5]
```diff
@@ -348,6 +356,232 @@ static int xocl_context_ioctl(struct xocl_dev *xdev, void *data,
...
+        return ret;
+}
+
+int xocl_close_cu_context(struct xocl_dev *xdev, struct drm_file *filp,
+		struct drm_xocl_close_cu_ctx *drm_cu_args)
+{
+	struct kds_client *client = filp->driver_priv;
+	struct kds_client_hw_ctx *hw_ctx = NULL;
+	struct kds_client_cu_ctx *cu_ctx = NULL;
+        struct kds_client_cu_info cu_info = { 0 }; // Local Variable
```
Comment (rozumx) [inline src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c:551]:
```suggestion
        struct kds_client_cu_info cu_info = { }; // Local Variable  
```

File: src/runtime_src/core/pcie/linux/device_linux.cpp [hunk 6]
```diff
@@ -356,6 +356,54 @@ struct sdm_sensor_info
...
+      throw xrt_core::query::sysfs_error(errmsg);
+
+    result_type xclbin_data;
+    // xclbin_uuid e.g.
+    // 0 <uuid_slot_0>
+    // 1 <uuid_slot_1>
+    for (auto& line : xclbin_info) {
+      boost::char_separator<char> sep(" ");
+      tokenizer tokens(line, sep);
+      slot_info data = { 0 };
```
Comment (rozumx) [inline src/runtime_src/core/pcie/linux/device_linux.cpp:383]:
```suggestion
      slot_info data = { };
```

File: tests/validate/verify_hw_ctx_test/src/host.cpp [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+            return EXIT_SUCCESS;
+        }
+    }
+
+    if (!infile.good()) {
+        std::cout << "\nNOT SUPPORTED" << std::endl;
+        return EOPNOTSUPP;
+    }
+
+    auto num_devices = xrt::system::enumerate_devices();
```
Comment (rozumx) [inline tests/validate/verify_hw_ctx_test/src/host.cpp:78]: Being that this is reference code.  The more comments regarding the flow the better.

**Review by rozumx — APPROVED** [review]
Minor requests.  Nothing to prevent approval.

**Comment by saifuddin-xilinx** [comment]
retest this please

File: src/runtime_src/core/common/drv/include/kds_core.h [hunk 8]
```diff
@@ -210,6 +211,22 @@ int kds_add_context(struct kds_sched *kds, struct kds_client *client,
 		    struct kds_client_cu_ctx *cu_ctx);
 int kds_del_context(struct kds_sched *kds, struct kds_client *client,
 		    struct kds_client_cu_ctx *cu_ctx);
+/* End of legacy context functions */
+
+/* Start of hw context functions */
```
Comment (mamin506) [inline src/runtime_src/core/common/drv/include/kds_core.h:216]: I will suggest you have a new set of APIs in this section.
If some APIs are implemented in the legacy functions, move the implementation to new functions. In legacy function, call the new one.

In clean up stage, delete old data structure and old APIs in the header file, then go fix all of the compile issue.

File: src/runtime_src/core/common/drv/include/kds_client.h [hunk 9]
```diff
@@ -87,8 +103,14 @@ struct kds_client {
...
-	struct kds_client_ctx    *ctx;
-	struct list_head          ctx_list;
+	struct kds_client_ctx    	*ctx;
+
+	/* To suppot ZOCL  multiple PL support */
+	struct list_head          	ctx_list;
+
+	/* To suppot multiple hw context */
+	struct list_head          	hw_ctx_list;
+	uint32_t 		 	next_avail_hw_ctx;
```
Comment (mamin506) [inline src/runtime_src/core/common/drv/include/kds_client.h:113]: In the use case, it looks like this is use for hw_ctx id. To make the name clear, how about name it as hw_ctx_id_counter or next_hw_ctx_id?

**Review by maxzhen — APPROVED** [review]

File: src/runtime_src/core/common/drv/include/kds_client.h [hunk 10]
```diff
@@ -28,11 +28,12 @@
 /* Multiple CU context can be active under a single KDS client Context.
  */
 struct kds_client_cu_ctx {
-	struct kds_client_ctx		*ctx;
 	u32				cu_idx;
 	u32		  		cu_domain;
 	u32				flags;
 	u32				ref_cnt;
+	struct kds_client_ctx		*ctx;
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/drv/include/kds_client.h:35]: This is for legacy context. Will be removed once hw context is fully functional.

File: src/runtime_src/core/common/drv/include/kds_client.h [hunk 11]
```diff
@@ -41,16 +42,31 @@ struct kds_client_cu_info {
...
 struct kds_client_ctx {
-	/* To support multiple context for multislot case */
+	void				*xclbin_id;
+	/* To support multiple CU context */
+	struct list_head		cu_ctx_list;
+
+	/* To support zocl multiple PL slot case */
 	struct list_head		link;
+	u32				slot_idx;
+};
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/drv/include/kds_client.h:59]: Yes. Will be removed in future.

File: src/runtime_src/core/common/drv/include/kds_client.h [hunk 12]
```diff
@@ -41,16 +42,31 @@ struct kds_client_cu_info {
...
+ */
+struct kds_client_hw_ctx {
+	uint32_t 			hw_ctx_idx;
 	void				*xclbin_id;
 	u32				slot_idx;
+	/* To support multiple context for multislot case */
+	struct list_head		link;
 	/* To support multiple CU context */
 	struct list_head		cu_ctx_list;
 };
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/drv/include/kds_client.h:72]: I have sent you a mail for the details. Thanks.

File: src/runtime_src/core/common/drv/include/kds_client.h (hunk 9 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/drv/include/kds_client.h:113]: Thanks for the suggestion. Updated accordingly.

File: src/runtime_src/core/common/drv/include/kds_core.h (hunk 8 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/drv/include/kds_core.h:216]: Will take care in Future PRs. Thanks for the suggestion.

File: src/runtime_src/core/common/drv/kds_core.c [hunk 13]
```diff
@@ -1071,20 +1142,61 @@ _kds_fini_client(struct kds_sched *kds, struct kds_client *client,
...
+
+	kds_info(client, "Client pid(%d) has open context for %d slot",
+			pid_nr(client->pid), hw_ctx->slot_idx);
+
+	mutex_lock(&client->lock);
+	/* Traverse through all the context and free them up */
+	list_for_each_entry_safe(cu_ctx, next, &hw_ctx->cu_ctx_list, link) {
+		kds_info(client, "Removing CU Domain[%d] CU Index [%d]", cu_ctx->cu_domain,
+				cu_ctx->cu_idx);
+		kds_del_context(kds, client, cu_ctx);
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/drv/kds_core.c:1168]: Updated. Thanks

File: src/runtime_src/core/common/drv/kds_core.c [hunk 14]
```diff
@@ -1288,6 +1418,156 @@ int kds_del_context(struct kds_sched *kds, struct kds_client *client,
...
+	/* Initialize the hw context here */
+	hw_ctx->hw_ctx_idx = client->next_avail_hw_ctx;
+	hw_ctx->slot_idx = slot_id;
+	hw_ctx->xclbin_id = xclbin_id;
+
+	/* Update next available hw context for this client */
+	++client->next_avail_hw_ctx;
+
+	/* Multiple CU context can be active. Initializing CU context list */
+	INIT_LIST_HEAD(&hw_ctx->cu_ctx_list);
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/drv/kds_core.c:1542]: Done. Thanks

File: src/runtime_src/core/common/drv/kds_core.c [hunk 15]
```diff
@@ -1288,6 +1418,156 @@ int kds_del_context(struct kds_sched *kds, struct kds_client *client,
...
+
+        /* Add this hw context to Client Context list */
+        list_add_tail(&hw_ctx->link, &client->hw_ctx_list);
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/drv/kds_core.c:1545]: Done. Thanks

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c [hunk 16]
```diff
@@ -225,19 +229,24 @@ static int xocl_add_context(struct xocl_dev *xdev, struct kds_client *client,
 		ret = -EINVAL;
 		goto out1;
 	}
-	 
+
 	ret = kds_add_context(&XDEV(xdev)->kds, client, cu_ctx);
-	if (ret)
+	if (ret) {
 		kds_free_cu_ctx(client, cu_ctx);
+		goto out1;
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c:236]: We are already in failure path.  Return value is of no use by now.

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c [hunk 17]
```diff
@@ -348,6 +356,232 @@ static int xocl_context_ioctl(struct xocl_dev *xdev, void *data,
...
+	hw_ctx = kds_alloc_hw_ctx(client, xclbin_id, slot_id);
+	if (!hw_ctx) {
+		ret = -EINVAL;
+		goto error_out;
+	}
+
+	/* Lock the bitstream. Unlock the same in destroy context */
+	ret = xocl_icap_lock_bitstream(xdev, xclbin_id);
+	if (ret) {
+		kds_free_hw_ctx(client, hw_ctx);
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c:385]: We are already in failure path. Return value is of no use by now.

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c [hunk 18]
```diff
@@ -348,6 +356,232 @@ static int xocl_context_ioctl(struct xocl_dev *xdev, void *data,
...
+                        continue;
+
+                if ((xcu->info.slot_idx == slot_hndl) &&
+                                (!strcmp(xcu->info.kname, kname)) &&
+                                (!strcmp(xcu->info.iname, iname))) {
+                        cu_info->cu_domain = DOMAIN_PS;
+                        cu_info->cu_idx = i;
+                        goto done;
+                }
+        }
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c:468]: 1. In this function we can't get the pointer based on domain as that is not yet discovered. 
Hence the search will be sequential. We can reduce some amount of code but not sure anything much better solution. 
2. I believe, That is the assumption that cu name can't be same. For today's solution also there is no way to the userspace to distinguish the cu index if the name is same.  We can ignore that case.

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c [hunk 19]
```diff
@@ -348,6 +356,232 @@ static int xocl_context_ioctl(struct xocl_dev *xdev, void *data,
...
+        if (cu_args->flags == XOCL_CTX_EXCLUSIVE)
+                cu_info->flags = CU_CTX_EXCLUSIVE;
+        else
+                cu_info->flags = CU_CTX_SHARED;
+
+        return 0;
+}
+
+static inline void
+xocl_hw_ctx_to_info(struct drm_xocl_close_cu_ctx *args,
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c:483]: Updated

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c [hunk 20]
```diff
@@ -348,6 +356,232 @@ static int xocl_context_ioctl(struct xocl_dev *xdev, void *data,
...
+{
+	struct kds_client *client = filp->driver_priv;
+	struct kds_client_hw_ctx *hw_ctx = NULL;
+	struct kds_client_cu_ctx *cu_ctx = NULL;
+	struct kds_client_cu_info cu_info = { 0 }; // Local Variable
+	int ret = 0;
+
+        /* Bitstream is locked. No one could load a new one
+         * until this client close all of the contexts.
+         */
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c:502]: True. Updated.

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c [hunk 21]
```diff
@@ -348,6 +356,232 @@ static int xocl_context_ioctl(struct xocl_dev *xdev, void *data,
...
+        mutex_lock(&client->lock);
+
+	/* Try to find the corresponding HW context */
+        hw_ctx = kds_get_hw_ctx_by_id(client, drm_cu_args->hw_context);
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c:506]: Sure. Thanks

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c [hunk 22]
```diff
@@ -348,6 +356,232 @@ static int xocl_context_ioctl(struct xocl_dev *xdev, void *data,
...
+        if (!hw_ctx) {
+                userpf_err(xdev, "No valid HW context is open");
+		mutex_unlock(&client->lock);
+                return -EINVAL;
+        }
+
+	/* Bitstream is locked. No one could load a new one
+	 * until this client close all of the contexts.
+	 */
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c:515]: Sure. Thanks

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c [hunk 23]
```diff
@@ -348,6 +356,232 @@ static int xocl_context_ioctl(struct xocl_dev *xdev, void *data,
...
+	cu_ctx = kds_alloc_cu_hw_ctx(client, hw_ctx, &cu_info);
+	if (!cu_ctx) {
+		userpf_err(xdev, "Allocation of CU context failed");
+		mutex_unlock(&client->lock);
+		return -EINVAL;
+	}
+
+        ret = kds_add_context(&XDEV(xdev)->kds, client, cu_ctx);
+        if (ret) {
+                kds_free_cu_ctx(client, cu_ctx);
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c:533]: Already code is failing state.

File: src/runtime_src/core/common/api/context_mgr.cpp [hunk 24]
```diff
@@ -38,51 +28,57 @@ constexpr auto virtual_cu_idx = std::numeric_limits<unsigned int>::max();
...
+      {}
+    };
+
+    std::map<std::string, std::shared_ptr<ip>> m_nm2ip;
+    std::map<decltype(cuidx_type::index), std::shared_ptr<ip>> m_idx2ip;
+
+    const ip*
+    get(const std::string& ipname)
+    {
+      return m_nm2ip[ipname].get();
```
Comment (vboggara-xilinx) [inline src/runtime_src/core/common/api/context_mgr.cpp:54]: what if ipname doesn't exists in m_nm2ip map. please add code for ipname  check.

File: src/runtime_src/core/common/api/context_mgr.cpp [hunk 25]
```diff
@@ -38,51 +28,57 @@ constexpr auto virtual_cu_idx = std::numeric_limits<unsigned int>::max();
...
-  std::pair<cuidx_type, std::bitset<max_cus>*>
-  get_ipidx_ctx(xcl_hwctx_handle ctxhdl, const std::string& cuname)
-  {
-    try {
-      auto ipidx = m_device->get_cuidx(ctxhdl, cuname);
-      return {ipidx, get_ctx(ipidx)};
+    const ip*
+    get(cuidx_type ipidx)
+    {
+      return m_idx2ip[ipidx.index].get();
```
Comment (vboggara-xilinx) [inline src/runtime_src/core/common/api/context_mgr.cpp:60]: please add a check for ipidx in m_idx2ip map.

File: src/runtime_src/core/common/api/context_mgr.cpp [hunk 26]
```diff
@@ -38,51 +28,57 @@ constexpr auto virtual_cu_idx = std::numeric_limits<unsigned int>::max();
...
-    // translate ipidx to idx used in bitset.
-    // virtual cu is last entry in bitset.
-    // virtual cu is always in default domain 0.
-    return ipidx.index == virtual_cu_idx ? max_cus - 1 : ipidx.domain_index;
-  }
+    void
+    erase(cuidx_type ipidx)
+    {
+      const auto& cu = m_idx2ip.at(ipidx.index);
+      m_nm2ip.erase(cu->ipname);
```
Comment (vboggara-xilinx) [inline src/runtime_src/core/common/api/context_mgr.cpp:73]: please find the element in map before accessing it or deleting it.

**Review by vboggara-xilinx — APPROVED** [review]
added comments for minor changes.

**Review by mamin506 — APPROVED** [review]

**Comment by kbhardwa-xilinx** [comment]
retest this please

File: src/runtime_src/core/common/config_reader.h (hunk 1 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/config_reader.h:561]: Added comments.

File: src/runtime_src/core/common/drv/kds_core.c (hunk 2 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/drv/kds_core.c]: Removed. Thanks.

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c (hunk 4 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c]: Done.

File: tests/validate/verify_hw_ctx_test/src/host.cpp (hunk 7 above)
Comment (saifuddin-xilinx) [reply tests/validate/verify_hw_ctx_test/src/host.cpp:78]: Will add further comments once we will use the new application. Right now it's only a reference code.

**Comment by saifuddin-xilinx** [comment]
retest this please

**Comment by saifuddin-xilinx** [comment]
retest this please

**Comment by saifuddin-xilinx** [comment]
Created another one https://github.com/Xilinx/XRT/pull/7112 after rebased with latest master

**Pull request closed without merging**